org.eclipse.xtext.web/src/main/java/org/eclipse/xtext/web/server/model/PrecomputedServiceRegistry.java

```diff
@@ -52,6 +52,8 @@ public boolean removePrecomputedService(AbstractCachedService<? extends IService
 	 * document is changed.
 	 */
 	public Iterable<AbstractCachedService<? extends IServiceResult>> getPrecomputedServices() {
-		return precomputedServices;
+		synchronized (precomputedServices) {
```

Comment:
What do you think about turning the field type for `precomputedServices` into a `CopyOnWriteArraySet` and remove all the synchronization? Or it could make sense to use `Collections.newSetFromMap(new ConcurrentHashMap<>())` .
Both approaches would save us from the synchronization blocks in this class.

Reply:
I think we should use `CopyOnWriteArraySet` here. The frequency of the calls to `getPrecomputedServices()` seems much higher than `add/removePrecomputedService()`.

Please correct me if am wrong with my assertion.

Reply:
:+1: